### PR TITLE
Add : Dockerized the whole project for easy setup and development

### PR DESCRIPTION
**Pull request by abbasidaniyal**
I have Dockerized the backend part of the code. 
Will work on dockerizing the whole project using docker-compose next!
WIP for : #3

### CODE REVIEW

**Comment by abbasidaniyal** [comment]
I have dockerized the whole project for development so that it becomes easy for people to get started with the project.
Also add the required steps in the README.md
If I missed out anything, do let me know.

Fixes : #3

**Comment by srish** [comment]
Thank you @abbasidaniyal for working on this! I will review your PR in a day or so.

File: README.md [hunk 1]
```diff
@@ -21,8 +21,19 @@ To view a community members’ contributions, provide their full name, Gerrit us
 There is also a feature to upload a list of usernames in a CSV format and view contributions for several community members altogether (this feature is still in experimental mode). 
 
 ## Getting started
-* First, clone the repo `git clone https://github.com/wikimedia/WikiContrib.git`
-* This tool has two different components: Frontend & Backend. You will have to set up both for the tool to be fully functional, and the instructions do so are here:
+- First, clone the repo `git clone https://github.com/wikimedia/WikiContrib.git`
+
+#### Using Docker Compose
```
Comment (srish) [inline README.md]: These instructions are for running the project, right? And, for when one has already followed frontend and backend Docker instructions. This wasn't clear to me immediately, and maybe it should be made clear here, and links to separate installation guides should be added.

File: backend/WikiContrib/Install.md [hunk 2]
```diff
@@ -6,6 +6,16 @@ The present doc deals about installing and running `backend`. If you go inside `
 
 # Steps to setup server locally
 
+## Using Docker 
+
+1. Install Docker in your system (https://docs.docker.com/install/)
```
Comment (srish) [inline backend/WikiContrib/Install.md]: Maybe for this step in both `backend/../Install.md` & `frontend/../Install.md/` add something like:
* One should ensure that the Docker has been successfully installed and only then should proceed. 
* This step should be executed once for backend & frontend.

File: backend/WikiContrib/Install.md [hunk 3]
```diff
@@ -6,6 +6,16 @@ The present doc deals about installing and running `backend`. If you go inside `
...
+2. Get to the  Dockerfile directory : `cd backend/WikiContrib`  
+3. Run `docker build .` to build an image.
+4. Run `docker  run -td  -p 8000:8000  <image-id>` to run the container.
```
Comment (srish) [inline backend/WikiContrib/Install.md]: What is image-id here? May be add some more info here and how to obtain it from the results of the previous step.

**Comment by abbasidaniyal** [comment]
Thank you for the review.
I'll try to make the documentation more user friendly!
Will make the required changes and push by tonight!

File: backend/WikiContrib/Install.md (hunk 3 above)
Comment (abbasidaniyal) [reply backend/WikiContrib/Install.md]: Noted!

File: README.md (hunk 1 above)
Comment (abbasidaniyal) [reply README.md]: Sure on it!

**Comment by abbasidaniyal** [comment]
> @abbasidaniyal Thank you so much for your work on this! I've left minor comments and then I've one error to report which I'm sharing below:
> 
> When I do `docker build .` for frontend, I get this as the final result instead of an image id:
> 
> ```
> Step 4/5 : RUN npm install --silent
>  ---> Running in dead87220933
> [23:45:03] Using gulpfile /app/node_modules/semantic-ui/gulpfile.js
> [23:45:03] Starting 'install'...
> ...
> [23:45:04] Starting 'run setup'...
> ? Set-up Semantic UI (Use arrow keys)
> ❯ Automatic (Use default locations and all components) 
>   Express (Set components and output folder) 
>   Custom (Customize all src/dist values) The command '/bin/sh -c npm install --silent' returned a non-zero code: 130
> ```
> 
> This may also have something to do with my node install as well, not sure. So, if you can address these comments, that'd be sweet!



>  Set-up Semantic UI (Use arrow keys)

Hi
Yes turns out, inorder to escape the interactive installation, we had to remove the semantic.json file from gitignore. It works fine now. Have a look and let me know if I missed out something!

**Comment by srish** [comment]
@abbasidaniyal Pretty close! Now, I am getting the following error when I run `docker-compose up`:
>backend-end_1  | /usr/local/lib/python3.6/site-packages/environ/environ.py:630: UserWarning: /code/WikiContrib/.env doesn't exist - if you're not configuring your environment separately, create one.
backend-end_1  |   "environment separately, create one." % env_file)

I put the .env file inside `backend/WikiContrib/WikiContrib/` still I'm getting the same error.

**Comment by abbasidaniyal** [comment]
> @abbasidaniyal Pretty close! Now, I am getting the following error when I run `docker-compose up`:
> 
> > backend-end_1  | /usr/local/lib/python3.6/site-packages/environ/environ.py:630: UserWarning: /code/WikiContrib/.env doesn't exist - if you're not configuring your environment separately, create one.
> > backend-end_1  |   "environment separately, create one." % env_file)
> 
> I put the .env file inside `backend/WikiContrib/WikiContrib/` still I'm getting the same error.


Did you put the .env file before running `docker-compose up --build` ?

**Comment by abbasidaniyal** [comment]
> > @abbasidaniyal Pretty close! Now, I am getting the following error when I run `docker-compose up`:
> > > backend-end_1  | /usr/local/lib/python3.6/site-packages/environ/environ.py:630: UserWarning: /code/WikiContrib/.env doesn't exist - if you're not configuring your environment separately, create one.
> > > backend-end_1  |   "environment separately, create one." % env_file)
> > 
> > 
> > I put the .env file inside `backend/WikiContrib/WikiContrib/` still I'm getting the same error.
> 
> Did you put the .env file before running `docker-compose up --build` ?

I have updated the instructions. Check and let me know if it works now.

**Comment by srish** [comment]
@abbasidaniyal Almost there! This is coming along nicely :) I have a few questions/comments, and then I would be happy to merge your changes as I think it is tremendous progress. If there are any remaining issues, they can be filed separately. 

So, now it seems like the frontend and backend is getting compiled without any errors, and here is what I get when I run docker-compose up:

```
wmf1699:WikiContrib ssethi$ docker-compose up
Starting wikicontrib_backend-end_1 ... done
Starting wikicontrib_front-end_1   ... done
Attaching to wikicontrib_front-end_1, wikicontrib_backend-end_1
front-end_1    | 
front-end_1    | > wikicontrib-frontend@0.1.0 start /app
front-end_1    | > react-scripts start
front-end_1    | 
backend-end_1  | Watching for file changes with StatReloader
backend-end_1  | Performing system checks...
backend-end_1  | 
backend-end_1  | System check identified no issues (0 silenced).
backend-end_1  | 
backend-end_1  | You have 21 unapplied migration(s). Your project may not work properly until you apply the migrations for app(s): admin, auth, contenttypes, query, result, sessions.
backend-end_1  | Run 'python manage.py migrate' to apply them.
backend-end_1  | March 13, 2020 - 22:59:31
backend-end_1  | Django version 2.2.10, using settings 'WikiContrib.settings'
backend-end_1  | Starting development server at http://0.0.0.0:8000/
backend-end_1  | Quit the server with CONTROL-C.
front-end_1    | Starting the development server...
front-end_1    | 
front-end_1    | Browserslist: caniuse-lite is outdated. Please run next command `npm update`
front-end_1    | Compiled successfully!
front-end_1    | 
front-end_1    | You can now view wikicontrib-frontend in the browser.
front-end_1    | 
front-end_1    |   Local:            http://localhost:3000/
front-end_1    |   On Your Network:  http://172.18.0.3:3000/
front-end_1    | 
front-end_1    | Note that the development build is not optimized.
front-end_1    | To create a production build, use npm run build.
front-end_1    | 
```

First, the instruction in the result above to apply migrations, seems odd. I have no experience with Docker, still I'm wondering if it should be part of it as well (but can be totally filed as separate task if it makes sense). 
Second, perhaps the `BASE_API_URL` should be with have a forward slash towards the end. But, even with that I get the error below when I try to search for a user:
```
backend-end_1  | Internal Server Error: /query/add/user/
backend-end_1  | Traceback (most recent call last):
backend-end_1  |   File "/usr/local/lib/python3.6/site-packages/django/db/backends/utils.py", line 84, in _execute
backend-end_1  |     return self.cursor.execute(sql, params)
backend-end_1  |   File "/usr/local/lib/python3.6/site-packages/django/db/backends/sqlite3/base.py", line 383, in execute
backend-end_1  |     return Database.Cursor.execute(self, query, params)
backend-end_1  | sqlite3.OperationalError: no such table: query_query
backend-end_1  | 
backend-end_1  | The above exception was the direct cause of the following exception:
backend-end_1  | 
backend-end_1  | Traceback (most recent call last):
backend-end_1  |   File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 34, in inner
backend-end_1  |     response = get_response(request)
backend-end_1  |   File "/usr/local/lib/python3.6/site-packages/django/core/handlers/base.py", line 115, in _get_response
backend-end_1  |     response = self.process_exception_by_middleware(e, request)
backend-end_1  |   File "/usr/local/lib/python3.6/site-packages/django/core/handlers/base.py", line 113, in _get_response
backend-end_1  |     response = wrapped_callback(request, *callback_args, **callback_kwargs)
backend-end_1  |   File "/usr/local/lib/python3.6/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
backend-end_1  |     return view_func(*args, **kwargs)
backend-end_1  |   File "/usr/local/lib/python3.6/site-packages/django/views/generic/base.py", line 71, in view
backend-end_1  |     return self.dispatch(request, *args, **kwargs)
backend-end_1  |   File "/usr/local/lib/python3.6/site-packages/rest_framework/views.py", line 495, in dispatch
backend-end_1  |     response = self.handle_exception(exc)
backend-end_1  |   File "/usr/local/lib/python3.6/site-packages/rest_framework/views.py", line 455, in handle_exception
backend-end_1  |     self.raise_uncaught_exception(exc)
backend-end_1  |   File "/usr/local/lib/python3.6/site-packages/rest_framework/views.py", line 492, in dispatch
backend-end_1  |     response = handler(request, *args, **kwargs)
backend-end_1  |   File "/code/query/views.py", line 132, in post
backend-end_1  |     request.data['hash_code'] = create_hash()
backend-end_1  |   File "/code/query/views.py", line 27, in create_hash
backend-end_1  |     while Query.objects.filter(hash_code=hash_code).exists():
backend-end_1  |   File "/usr/local/lib/python3.6/site-packages/django/db/models/query.py", line 766, in exists
backend-end_1  |     return self.query.has_results(using=self.db)
backend-end_1  |   File "/usr/local/lib/python3.6/site-packages/django/db/models/sql/query.py", line 522, in has_results
backend-end_1  |     return compiler.has_results()
backend-end_1  |   File "/usr/local/lib/python3.6/site-packages/django/db/models/sql/compiler.py", line 1103, in has_results
backend-end_1  |     return bool(self.execute_sql(SINGLE))
backend-end_1  |   File "/usr/local/lib/python3.6/site-packages/django/db/models/sql/compiler.py", line 1133, in execute_sql
backend-end_1  |     cursor.execute(sql, params)
backend-end_1  |   File "/usr/local/lib/python3.6/site-packages/django/db/backends/utils.py", line 99, in execute
backend-end_1  |     return super().execute(sql, params)
backend-end_1  |   File "/usr/local/lib/python3.6/site-packages/django/db/backends/utils.py", line 67, in execute
backend-end_1  |     return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
backend-end_1  |   File "/usr/local/lib/python3.6/site-packages/django/db/backends/utils.py", line 76, in _execute_with_wrappers
backend-end_1  |     return executor(sql, params, many, context)
backend-end_1  |   File "/usr/local/lib/python3.6/site-packages/django/db/backends/utils.py", line 84, in _execute
backend-end_1  |     return self.cursor.execute(sql, params)
backend-end_1  |   File "/usr/local/lib/python3.6/site-packages/django/db/utils.py", line 89, in __exit__
backend-end_1  |     raise dj_exc_value.with_traceback(traceback) from exc_value
backend-end_1  |   File "/usr/local/lib/python3.6/site-packages/django/db/backends/utils.py", line 84, in _execute
backend-end_1  |     return self.cursor.execute(sql, params)
backend-end_1  |   File "/usr/local/lib/python3.6/site-packages/django/db/backends/sqlite3/base.py", line 383, in execute
backend-end_1  |     return Database.Cursor.execute(self, query, params)
backend-end_1  | django.db.utils.OperationalError: no such table: query_query
backend-end_1  | [13/Mar/2020 23:05:11] "POST /query/add/user/ HTTP/1.1" 500 16242
```
Thoughts?

**Comment by abbasidaniyal** [comment]
I have made a small change which will take care of the migrations problem. Also, I think the second problem is due to the first one where the migrations have not been applied.

**Comment by srish** [comment]
@abbasidaniyal That was quick! Thanks a lot and it works like a charm :)

There was a small error that I encountered while building backend due to tokens like `(` in a variable `SECRET_KEY` in `.env` that maybe the script couldn't parse:
> Running in 761b2b645ad7
/code/WikiContrib/.env: line 15: command not found
The command '/bin/bash -c source /code/WikiContrib/.env' returned a non-zero code: 127

But, as `SECRET_KEY` is production specific, it is totally fine. I cleared the token value and I was able to get through it.

**Comment by abbasidaniyal** [comment]
That's great! Really happy with my contribution. Looking forward to many more! 😁

**Pull request merged**